File: src/Networking/EngineStatus.cpp
#include "EngineStatus.hpp"

using namespace GForce::Networking;

EngineStatus::EngineStatus(bool powerAmplifierReleased, bool inverterReady, bool paDataReady,
                                               int currentIntegratorSet, int currentParameterSet, bool malfunction,
                                               bool switchRightActive, bool switchLeftActive) : powerAmplifierReleased(
        powerAmplifierReleased), inverterReady(inverterReady), inputDataReady(paDataReady), currentIntegratorSet(
        currentIntegratorSet), currentParameterSet(currentParameterSet), malfunction(malfunction), switchRightActive(
        switchRightActive), switchLeftActive(switchLeftActive) {}


EngineStatus *EngineStatus::clone()
{
    return new EngineStatus(powerAmplifierReleased, inverterReady, inputDataReady, currentIntegratorSet, currentParameterSet, malfunction, switchRightActive, switchLeftActive);
}

nlohmann::json EngineStatus::toJSON()
{
    nlohmann::json data = {
        {"powerAmplifierReleased", powerAmplifierReleased},
        {"inverterReady", inverterReady},
        {"inputDataReady", inputDataReady},
        {"currentIntegratorSet", currentIntegratorSet},
        {"currentParameterSet", currentParameterSet},
        {"malfunction", malfunction},
        {"switchRightActive", switchRightActive},
        {"switchLeftActive", switchLeftActive},
    };

    return data;
}

bool EngineStatus::isPowerAmplifierReleased() const {
    return powerAmplifierReleased;
}

bool EngineStatus::isInverterReady() const {
    return inverterReady;
}

bool EngineStatus::isInputDataReady() const {
    return inputDataReady;
}

int EngineStatus::getCurrentIntegratorSet() const {
    return currentIntegratorSet;
}

int EngineStatus::getCurrentParameterSet() const {
    return currentParameterSet;
}

bool EngineStatus::isMalfunction() const {
    return malfunction;
}

bool EngineStatus::isSwitchRightActive() const {
    return switchRightActive;
}

bool EngineStatus::isSwitchLeftActive() const {
    return switchLeftActive;
}

File: web/js/lib/Operation/EngineStatus.js
class EngineStatus 
{
    /**
     * @param {*} props
     */
    constructor(props)
    {
        /**
         * @type {boolean}
         */
        this.powerAmplifierReleased = props.powerAmplifierReleased;

        /**
         * @type {boolean}
         */
        this.inverterReady = props.inverterReady;

        /**
         * @type {boolean}
         */
        this.inputDataReady = props.inputDataReady;

        /**
         * @type {boolean}
         */
        this.malfunction = props.malfunction;

        /**
         * @type {boolean}
         */
        this.switchLeftActive = props.switchLeftActive;

        /**
         * @type {boolean}
         */
        this.switchRightActive = props.switchRightActive;

        /**
         * @type {number}
         */
        this.currentIntegratorSet = props.currentIntegratorSet;

        /**
         * @type {number}
         */
        this.currentParameterSet = props.currentParameterSet;
    }

    /**
     * @return {boolean}
     */
    isFullyReleased()
    {
        return this.powerAmplifierReleased && this.inverterReady && this.inputDataReady;
    }

    /**
     * @param {EngineStatus} status
     * @returns {boolean}
     */
    isEqual(status)
    {
        return this.powerAmplifierReleased === status.powerAmplifierReleased
            && this.inverterReady === status.inverterReady
            && this.inputDataReady === status.inputDataReady
            && this.malfunction === status.malfunction
            && this.switchLeftActive === status.switchLeftActive
            && this.switchRightActive === status.switchRightActive
            && this.currentIntegratorSet === status.currentIntegratorSet
            && this.currentParameterSet === status.currentParameterSet;
    }
}